pkg/mei/mkhiclient.go

```diff
@@ -0,0 +1,300 @@
+	"MEI_DEV_ID_MCC":        0x4B70,
+	"MEI_DEV_ID_MCC_4":      0x4B75,
+}
+
+// MKHIClient is a client to send MKHI commands via MEI.
+type MKHIClient struct {
+	MEI              MEI
+	clientProperties *ClientProperties
+}
+
+// MKHI command groups, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L22
+const (
+	MKHIGroupIDCbm    = 0x0
+	MKHIGroupIDHmrfpo = 0x5
+	MKHIGroupIDGen    = 0xff
+)
+
+// MKHI HMRFPO command IDs, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L33
+const (
+	MKHIHmrfpoEnable    = 0x1
+	MKHIHmrfpoGetStatus = 0x3
+)
+
+// pciAddressToInt converts the last part of a PCI address string to integer,
+// e.g. 0000:00:16.0 will return 0x16 .
+func pciAddressToInt(pciaddr string) (int, error) {
+	parts := strings.Split(pciaddr, ":")
+	if len(parts) != 3 {
+		return 0, fmt.Errorf("wrong PCI address string '%s': want 3 colon-separated groups, got %d", pciaddr, len(parts))
+	}
+	parts = strings.Split(parts[2], ".")
+	if len(parts) != 2 {
+		return 0, fmt.Errorf("wrong PCI address sub-component '%s': want 2 dot-separated groups, got %d", pciaddr, len(parts))
+	}
+	a, err := strconv.ParseInt(parts[0], 16, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse hex string '%s': %w", parts[0], err)
+	}
+	return int(a), nil
+}
+
+// Open opens an MKHI client connection.
+func (m *MKHIClient) Open(meiPath string) error {
+	cp, err := m.MEI.Open(meiPath, MKHIGuid)
+	if err != nil {
+		return err
+	}
+	m.clientProperties = cp
+	return nil
+}
+
+// Close closes an MKHI client connection.
+func (m *MKHIClient) Close() error {
+	m.clientProperties = nil
+	return m.MEI.Close()
+}
+
+// IsHmrfpoEnableAllowed queries whether the HMRFPO enable is allowed.
+func (m MKHIClient) IsHmrfpoEnableAllowed() (bool, error) {
+	/* This is a reimplementation of cse_is_hmrfpo_enable_allowed from
+	 * coreboot/src/soc/intel/common/block/cse/cse.c . The below comment also
+	 * comes from that function.
+	 *
+	 * Allow sending HMRFPO ENABLE command only if:
+	 *  - CSE's current working state is Normal and current operation mode is Normal
+	 *  - (or) cse's current working state is normal and current operation mode is
+	 *    Soft Temp Disable if CSE's Firmware SKU is Custom
+	 *
+	 */
+	meiDev, err := GetMeiPciDevice()
+	if err != nil {
+		return false, fmt.Errorf("failed to get PCI MEI device: %w", err)
+	}
+	log.Printf("MEI Device found: %s", meiDev)
+	intelBusID := 0
+	intelMEDeviceID, err := pciAddressToInt(meiDev.Address)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse MEI product ID: %w", err)
+	}
+	// usually 0. Any better way to discover this dynamically?
+	intelMEDevFn := 0
+
+	// check that CSE's current working state is normal
+	cs, err := txtsuite.PCIReadConfig32(intelBusID, int(intelMEDeviceID), intelMEDevFn, 0x40 /* PCI_ME_HFSTS1 */)
```

Comment:
If you're just using txtsuite for pci, perhaps try pkg/pci locally? It may be missing some functionality, but it's already in our set of dependencies and would probably be easy to extend for what you need

Reply:
yeah good point, I'll try pkg/pci (I didn't realize it's there)

Reply:
done